File: Managed Source/Arsenal.ImageMounter.Devio/Server/GenericProviders/DevioProviderDLLWrapperBase.vb

''''' DevioProviderDLLWrapperBase.vb
''''' Proxy provider that implements devio proxy service with an unmanaged DLL written
''''' for use with devio.exe command line tool.
''''' 
''''' Copyright (c) 2012-2021, Arsenal Consulting, Inc. (d/b/a Arsenal Recon) <https://www.ArsenalRecon.com>
''''' This source code and API are available under the terms of the Affero General Public
''''' License v3.
'''''
''''' Please see LICENSE.txt for full license terms, including the availability of
''''' proprietary exceptions.
''''' Questions, comments, or requests for clarification: https://ArsenalRecon.com/contact/
'''''

Imports System.ComponentModel
Imports System.Diagnostics.CodeAnalysis
Imports System.IO
Imports System.Runtime.InteropServices
Imports Microsoft.Win32.SafeHandles

Namespace Server.GenericProviders

    ''' <summary>
    ''' Class that implements <see>IDevioProvider</see> interface with an unmanaged DLL
    ''' written for use with devio.exe command line tool.
    ''' object as storage backend.
    ''' </summary>
    Public MustInherit Class DevioProviderDLLWrapperBase
        Inherits DevioProviderUnmanagedBase

#Region "SafeHandle"
        Public Class SafeDevioProviderDLLHandle
            Inherits SafeHandleZeroOrMinusOneIsInvalid

            Protected Friend Property DLLClose As DLLCloseMethod

            Public Sub New(handle As IntPtr, ownsHandle As Boolean)
                MyBase.New(ownsHandle)

                SetHandle(handle)
            End Sub

            Protected Sub New()
                MyBase.New(True)

            End Sub

            Protected Overrides Function ReleaseHandle() As Boolean
                If _DLLClose Is Nothing Then
                    Return True
                End If
                Return _DLLClose(handle) <> 0
            End Function
        End Class
#End Region

        Protected Sub New(open As DLLOpenMethod, filename As String, [readOnly] As Boolean)
            Me.New(open, filename, [readOnly], Nothing)

        End Sub

        Protected Sub New(open As DLLOpenMethod, filename As String, [readOnly] As Boolean, get_last_error As Func(Of Exception))

            If open Is Nothing Then
                Throw New ArgumentNullException(NameOf(open))
            End If

            Dim dll_close As DLLCloseMethod = Nothing

            _SafeHandle = open(filename, [readOnly], _DLLRead, _DLLWrite, dll_close, _Length)

            If _SafeHandle.IsInvalid OrElse _SafeHandle.IsClosed Then
                Throw New IOException($"Error opening '{filename}'", If(get_last_error?.Invoke(), New Win32Exception()))
            End If

            _SafeHandle.DLLClose = dll_close

            _CanWrite = Not [readOnly]

        End Sub

        Public ReadOnly Property SafeHandle As SafeDevioProviderDLLHandle

        Public Overrides ReadOnly Property Length As Long

        Public Overrides ReadOnly Property CanWrite As Boolean

        Public Overridable ReadOnly Property DLLRead As DLLReadWriteMethod

        Public Overridable ReadOnly Property DLLWrite As DLLReadWriteMethod

        Public Delegate Function DLLOpenMethod(<MarshalAs(UnmanagedType.LPStr), [In]> filename As String,
                                               <MarshalAs(UnmanagedType.Bool)> read_only As Boolean,
                                               <MarshalAs(UnmanagedType.FunctionPtr), Out> ByRef dllread As DLLReadWriteMethod,
                                               <MarshalAs(UnmanagedType.FunctionPtr), Out> ByRef dllwrite As DLLReadWriteMethod,
                                               <MarshalAs(UnmanagedType.FunctionPtr), Out> ByRef dllclose As DLLCloseMethod,
                                               <Out> ByRef size As Long) As SafeDevioProviderDLLHandle

        Public Delegate Function DLLReadWriteMethod(handle As SafeDevioProviderDLLHandle,
                                                    buffer As IntPtr,
                                                    size As Integer,
                                                    offset As Long) As Integer

        Public Delegate Function DLLCloseMethod(handle As IntPtr) As Integer

        Public Overrides Function Read(buffer As IntPtr, bufferoffset As Integer, count As Integer, fileoffset As Long) As Integer

            Return DLLRead(SafeHandle, buffer + bufferoffset, count, fileoffset)

        End Function

        Public Overrides Function Write(buffer As IntPtr, bufferoffset As Integer, count As Integer, fileoffset As Long) As Integer

            Return DLLWrite(SafeHandle, buffer + bufferoffset, count, fileoffset)

        End Function

        Protected Overrides Sub Dispose(disposing As Boolean)
            If disposing Then
                _SafeHandle?.Dispose()
            End If

            _SafeHandle = Nothing

            MyBase.Dispose(disposing)
        End Sub

    End Class

End Namespace

File: Managed Source/Arsenal.ImageMounter/Devio/Server/GenericProviders/DevioProviderDLLWrapperBase.cs
//  DevioProviderDLLWrapperBase.vb
//  Proxy provider that implements devio proxy service with an unmanaged DLL written
//  for use with devio.exe command line tool.
//  
//  Copyright (c) 2012-2025, Arsenal Consulting, Inc. (d/b/a Arsenal Recon) <https://www.ArsenalRecon.com>
//  This source code and API are available under the terms of the Affero General Public
//  License v3.
// 
//  Please see LICENSE.txt for full license terms, including the availability of
//  proprietary exceptions.
//  Questions, comments, or requests for clarification: https://ArsenalRecon.com/contact/
// 

using Microsoft.Win32.SafeHandles;
using System;
using System.ComponentModel;
using System.IO;
using System.Runtime.InteropServices;

#pragma warning disable IDE0079 // Remove unnecessary suppression
#pragma warning disable CA1510 // Use ArgumentNullException throw helper

namespace Arsenal.ImageMounter.Devio.Server.GenericProviders;

/// <summary>
/// Class that implements <see>IDevioProvider</see> interface with an unmanaged DLL
/// written for use with devio.exe command line tool.
/// object as storage backend.
/// </summary>
public abstract class DevioProviderDLLWrapperBase : DevioProviderUnmanagedBase
{
    #region SafeHandle
    public class SafeDevioProviderDLLHandle : SafeHandleZeroOrMinusOneIsInvalid
    {
        protected internal unsafe delegate* unmanaged[Cdecl]<nint, int> DLLClose { get; set; }

        public SafeDevioProviderDLLHandle(nint handle, bool ownsHandle)
            : base(ownsHandle)
        {
            SetHandle(handle);
        }

        protected SafeDevioProviderDLLHandle()
            : base(true)
        {
        }

        protected override unsafe bool ReleaseHandle() => DLLClose is null || DLLClose(handle) != 0;
    }
    #endregion

    protected DevioProviderDLLWrapperBase(DLLOpenMethod open, string filename, bool readOnly)
        : this(open, filename, readOnly, null)
    {
    }

    protected DevioProviderDLLWrapperBase(DLLOpenMethodNIntRet open, string filename, bool readOnly)
        : this(open, filename, readOnly, null)
    {
    }

    protected unsafe DevioProviderDLLWrapperBase(DLLOpenMethodNIntRet open, string filename, bool readOnly, Func<Exception>? get_last_error)
        : this(GetDLLOpenWrapper(open), filename, readOnly, get_last_error)
    {
    }

    private static unsafe DLLOpenMethod GetDLLOpenWrapper(DLLOpenMethodNIntRet open)
    {
        SafeDevioProviderDLLHandle DLLOpenWrapper(string filename,
                                                  bool read_only,
                                                  out delegate* unmanaged[Cdecl]<nint, nint, int, long, int> dllread,
                                                  out delegate* unmanaged[Cdecl]<nint, nint, int, long, int> dllwrite,
                                                  out delegate* unmanaged[Cdecl]<nint, int> dllclose,
                                                  out long size)
        {
            var handle = open(filename, read_only, out dllread, out dllwrite, out dllclose, out size);
            return new(handle, ownsHandle: true);
        }

        return DLLOpenWrapper;
    }

    protected unsafe DevioProviderDLLWrapperBase(DLLOpenMethod open, string filename, bool readOnly, Func<Exception>? get_last_error)
    {
        if (open is null)
        {
            throw new ArgumentNullException(nameof(open));
        }

        SafeHandle = open(filename, readOnly, out var dllRead, out var dllWrite, out var dllClose, out var length);

        if (SafeHandle.IsInvalid || SafeHandle.IsClosed)
        {
            throw new IOException($"Error opening '{filename}'", (get_last_error?.Invoke()) ?? new Win32Exception());
        }

        DLLRead = dllRead;
        DLLWrite = dllWrite;
        Length = length;
        SafeHandle.DLLClose = dllClose;

        CanWrite = !readOnly;
    }

    public SafeDevioProviderDLLHandle SafeHandle { get; }

    public override long Length { get; }

    public override bool CanWrite { get; }

    protected internal unsafe delegate* unmanaged[Cdecl]<nint, nint, int, long, int> DLLRead { get; }

    protected internal unsafe delegate* unmanaged[Cdecl]<nint, nint, int, long, int> DLLWrite { get; }

    public unsafe delegate SafeDevioProviderDLLHandle DLLOpenMethod([MarshalAs(UnmanagedType.LPStr)] string filename,
                                                                    [MarshalAs(UnmanagedType.Bool)] bool read_only,
                                                                    out delegate* unmanaged[Cdecl]<nint, nint, int, long, int> dllread,
                                                                    out delegate* unmanaged[Cdecl]<nint, nint, int, long, int> dllwrite,
                                                                    out delegate* unmanaged[Cdecl]<nint, int> dllclose,
                                                                    out long size);

    public unsafe delegate nint DLLOpenMethodNIntRet([MarshalAs(UnmanagedType.LPStr)] string filename,
                                                     [MarshalAs(UnmanagedType.Bool)] bool read_only,
                                                     out delegate* unmanaged[Cdecl]<nint, nint, int, long, int> dllread,
                                                     out delegate* unmanaged[Cdecl]<nint, nint, int, long, int> dllwrite,
                                                     out delegate* unmanaged[Cdecl]<nint, int> dllclose,
                                                     out long size);

    public override unsafe int Read(nint buffer, int bufferoffset, int count, long fileoffset)
    {
        var ptrSuccess = false;

        try
        {
            SafeHandle.DangerousAddRef(ref ptrSuccess);

            return DLLRead(SafeHandle.DangerousGetHandle(), buffer + bufferoffset, count, fileoffset);
        }
        finally
        {
            if (ptrSuccess)
            {
                SafeHandle.DangerousRelease();
            }
        }
    }

    public override unsafe int Write(nint buffer, int bufferoffset, int count, long fileoffset)
    {
        var ptrSuccess = false;

        try
        {
            SafeHandle.DangerousAddRef(ref ptrSuccess);

            return DLLWrite(SafeHandle.DangerousGetHandle(), buffer + bufferoffset, count, fileoffset);
        }
        finally
        {
            if (ptrSuccess)
            {
                SafeHandle.DangerousRelease();
            }
        }
    }

    protected override void Dispose(bool disposing)
    {
        if (disposing)
        {
            SafeHandle?.Dispose();
        }

        base.Dispose(disposing);
    }
}
